### DISCUSSION TITLE
Continuity of the solution across material blocks

### DISCUSSION THREAD

**Discussion by gabburgio** [idaholab/moose#23511] [Q&A General] [answered]
Greetings

I am trying to solve a reaction-diffusion equation on a rectangular domain, which is itself made up of two rectangles (see attached image).
I have set the value of the solution on the boundary (Dirichlet BC) to be 5.
In the rectangle above the reaction coefficient is 100, in the one below it is 1:

```
  [Materials]
    [./example]
      type = ADPiecewiseConstantByBlockMaterial
      prop_name = 'reaction'
      subdomain_to_prop_value = 'zone1 1
                                 zone2 100'
    [../]
  []
```
The diffusion coefficient is 1 for both.
My question might be naive, but given that I selected the default linear Lagrange FE functions:

```
  [Variables]
    [./concentration]
    [../]
  []
```
shouldn't the solution itself be a continuous function, unlike my result?

![Dirichlet_DR](https://user-images.githubusercontent.com/119760805/220657447-8b98f0d5-99b7-411b-9a9f-3c927c2782bf.png)

I left the (very coarse) mesh from peacock in to show that, to my understanding, it is not a question of the function quickly going to zero, but rather a genuine discontinuity. (The result is essentially independent of refinement anyway)

I assumed that the different reaction coefficients would simply be reflected in the entries of the mass matrix associated with the reaction kernel (so larger entries on the side with the larger coefficient, and something in between for the FE functions associated with nodes on the boundary between the two rectangles), but I would say this is not the case, and the solution seems to be carried out independently on the two subdomains.

Is there a different way one should go about defining material zones? (I defined the two rectangles as physical groups in gmsh)
When dealing with discontinuous properties across material boundaries, should one always make use of Interface Kernels? (here I haven't)

Thank you for your time

Gabriele

**Comment by GiudGiud** [comment]
Hello

I would visualize it in paraview and check whether the results are really discontinuous. Peacock is less fully-featured than paraview, I dont think you can draw lines through the geometry to check that. 

You also seem to be plotting the flux, not the concentration. The flux being related to the first derivative, it s not necessarily continuous, and not with this variable type. 

As long as the mesh is conformal at the interface, you do not need to use interface kernels. Physics (kernels) that extends across both domains will naturally be solved at the interface. 

Guillaume

**Comment by gabburgio** [comment]
Sorry for the delay, about the two points:

The flux is actually the concentration variable. 
I am solving a very simplified neutron diffusion equation, which is a type of diffusion reaction equation, and I thought it would be easier to understand what I was talking about if I said diffusion reaction, but I did not change the name in the plot.

The same problem is present in Paraview
(the solution jumps from almost five to almost zero across the central boundary)
![fluximage](https://user-images.githubusercontent.com/119760805/223177355-ba422bfd-7cd4-4703-9983-a106949c1a8f.png)

Is it then reasonable to assume that the solution is carried out independently in the two subdomains?
If so, why is this happening?

Thanks again

**Reply by GiudGiud** [reply]
I would first draw line plots through the geometry to check for a discontinuity.

Then could you please attach the convergence log? To make sure it s converged. 

If the mesh is contiguous and you are not block restricting physics inappropriately, the solution should be computed with both domains coupled. You can check for this easily by making the two domains more similar then progressively modifying one and see the solution in the other one change

**Comment by gabburgio** [comment]
these are some line plots:

(vertical line)
![straightline](https://user-images.githubusercontent.com/119760805/223486243-22e4d198-7d71-4cae-8db9-689211f1374f.png)

(diagonal line)
![plotoverline](https://user-images.githubusercontent.com/119760805/223486350-9dcd7c16-f460-4677-99d0-7b6c6f126d16.png)

full domain plot:
![linepic](https://user-images.githubusercontent.com/119760805/223486511-e00bd6a5-4b66-49a2-808a-53aa1833de2c.png)


here is the convergence log:

```
Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   111
  Elems:                   170
  Num Subdomains:          2

Nonlinear System:
  Num DOFs:                111
  Num Local DOFs:          111
  Variables:               "flux" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             Preconditioned JFNK
  PETSc Preconditioner:    hypre boomeramg 

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored, and these tests should be re-golded.

 0 Nonlinear |R| = 3.311429e+01
      0 Linear |R| = 3.311429e+01
      1 Linear |R| = 4.168228e-01
      2 Linear |R| = 1.346020e-02
      3 Linear |R| = 5.631338e-04
      4 Linear |R| = 7.175489e-06
 1 Nonlinear |R| = 7.412694e-06
      0 Linear |R| = 7.412694e-06
      1 Linear |R| = 2.761694e-07
      2 Linear |R| = 3.649083e-09
      3 Linear |R| = 1.598945e-10
      4 Linear |R| = 3.063759e-12
 2 Nonlinear |R| = 3.264146e-12
 Solve Converged!

```
I also tried varying the reaction coefficient in one of the two domains and I could not discern a difference in the other one.

At the moment I believe that this discontinuity is due the gmsh mesh, since I produced it simply by creating two adjacent rectangular surfaces and assigning to each a physical group (zone1 and zone2).
This also results in the elements of the mesh all having either a vertex or a side on the boundary between the two domains, and none of them "crossing over".

I am investigating how to remedy this in gmsh, but if you have any advice (or any other meshing program to recommend) I would be very grateful

Thanks

Gabriele

**Reply by GiudGiud** [reply]
For such a simple mesh, you can use MOOSE meshgenerators
https://mooseframework.inl.gov/source/meshgenerators/CartesianMeshGenerator.html

I believe you are right about this now. Appending without merging two meshes will do this

**Comment by gabburgio** [comment]
As expected it works with the meshgenerator:


![meshgen](https://user-images.githubusercontent.com/119760805/223733537-1b5a9695-b56b-47da-83bf-7b47779f5abc.png)

However, I am going to need more general geometric capabilities in the future. 
I did manage to produce a single mesh on the two material zones in gmsh using compund surfaces, but then I found myself in the same situation as https://github.com/idaholab/moose/discussions/16247, since I get the error: 

`I don't believe that we can specify multiple subdomain or boundary ids for a given geometric entity`

Is using an "EqualValueBoundaryConstraint" on all internal boundaries, as suggested there, really the best approach to define several materials within a geometry?

Is there some other meshing software you would recommend for the purpose?

**Reply by GiudGiud** [reply]
There s still some issue with the mesh. It seems there's overlapping subdomain ids on a given element. This is fine in your mesher but not supported by MOOSE

**Comment by gabburgio** [comment, marked as answer]
As it turns out, the way to achieve this in gmsh is to use _boolean fragments_ on the surfaces/volumes one wants to join, e.g. _BooleanFragments{ Volume{1,2}; Delete; }{}_

**Reply by GiudGiud** [reply]
Thanks for reporting the solution!
